server/handlers/validation.go

```diff
@@ -125,34 +77,150 @@ func validateUpdate(gun string, updates []storage.MetaUpdate, store storage.Meta
 		if t, err = validateTargets(targetsRole, roles, kdb); err != nil {
 			logrus.Error("ErrBadTargets: ", err.Error())
-			return ErrBadTargets{msg: err.Error()}
+			return nil, ErrBadTargets{msg: err.Error()}
 		}
 		repo.SetTargets(targetsRole, t)
 	}
 	logrus.Debug("Successfully validated targets")
 
-	var oldSnap *data.SignedSnapshot
-	oldSnapJSON, err := store.GetCurrent(gun, snapshotRole)
-	if _, ok := err.(*storage.ErrNotFound); err != nil && !ok {
-		// problem with storage. No expectation we can
-		// write if we can't read so bail.
-		logrus.Error("error reading previous snapshot: ", err.Error())
-		return err
-	} else if err == nil {
-		oldSnap = &data.SignedSnapshot{}
-		if err := json.Unmarshal(oldSnapJSON, oldSnap); err != nil {
-			oldSnap = nil
+	if _, ok := roles[snapshotRole]; ok {
+		var oldSnap *data.SignedSnapshot
+		oldSnapJSON, err := store.GetCurrent(gun, snapshotRole)
+		if _, ok := err.(*storage.ErrNotFound); err != nil && !ok {
+			// problem with storage. No expectation we can
+			// write if we can't read so bail.
+			logrus.Error("error reading previous snapshot: ", err.Error())
+			return nil, err
+		} else if err == nil {
+			oldSnap = &data.SignedSnapshot{}
+			if err := json.Unmarshal(oldSnapJSON, oldSnap); err != nil {
+				oldSnap = nil
+			}
+		}
+
+		if err := validateSnapshot(snapshotRole, oldSnap, roles[snapshotRole], roles, kdb); err != nil {
+			logrus.Error("ErrBadSnapshot: ", err.Error())
+			return nil, ErrBadSnapshot{msg: err.Error()}
+		}
+		logrus.Debug("Successfully validated snapshot")
+	} else {
+		// Check:
+		//	 - we have a snapshot key
+		//   - it matches a snapshot key signed into the root.json
+		// Then:
+		//   - generate a new snapshot
+		//   - add it to the updates
+		//
+		// TODO: Assumptions have been made about what is loaded into the repo.
+		//       we need to ensure at a minimum the root and targets are loaded
+		//       in.
+		err := prepRepo(gun, repo, store)
+		if err != nil {
+			return nil, err
 		}
+		update, err := generateSnapshot(gun, kdb, repo, store)
+		if err != nil {
+			return nil, err
+		}
+		updates = append(updates, *update)
 	}
+	return updates, nil
+}
 
-	if err := validateSnapshot(snapshotRole, oldSnap, roles[snapshotRole], roles, kdb); err != nil {
-		logrus.Error("ErrBadSnapshot: ", err.Error())
-		return ErrBadSnapshot{msg: err.Error()}
+func prepRepo(gun string, repo *tuf.Repo, store storage.MetaStore) error {
+	if repo.Root == nil {
+		rootJSON, err := store.GetCurrent(gun, data.CanonicalRootRole)
+		if err != nil {
+			return fmt.Errorf("could not load repo for snapshot generation: %v", err)
+		}
+		root := &data.SignedRoot{}
+		err = json.Unmarshal(rootJSON, root)
+		if err != nil {
+			return fmt.Errorf("could not load repo for snapshot generation: %v", err)
+		}
+		repo.SetRoot(root)
+	}
+	if repo.Targets[data.CanonicalTargetsRole] == nil {
+		targetsJSON, err := store.GetCurrent(gun, data.CanonicalTargetsRole)
```

Comment:
Non-blocking - we seem to be getting the current role, unmarshalling it into a `data.Signed*` a lot.  I was wondering if it might be helpful to abstract these out to helper functions like `GetCurrentSignedt*`?